### DISCUSSION TITLE
Questions related to document layouts.

### DISCUSSION THREAD

**Discussion by no-vici** [Omikhleia/resilient.sile#37] [Q&A] [answered]
Hi again,

I've been diving into your code for ratio based layouts and found a few things that make my head spin. I hope you won't mind helping me clear the mist.

Parsing for the `division` layout in the `layoutParse` states that:

```
...
local number = SILE.parserBits.number
...
division = P("division")
    * (
      (ws * number * ws * number)
      + (ws * number)
      + (lpeg.Cc(9))
    ) / function(n, ratio)
    local layout = require("resilient.layouts.division")
    return layout({ n = n, ratio = ratio })
  end
```
I went into Lpeg document and found:

`lpeg.Cc ([value, ...])`: Creates a constant capture. This pattern matches the empty string and produces all given values as its captured values.

You used lpeg.Cc(9), so this returns 9. So what is the meaning of `(ws * number * ws * number) + (ws * number)`? I know the `ws` bit but not the `number` and the previous combination.

The `function(n, ratio)` returns `layout({ n = n, ratio = ratio })`, so n = 9 and what is the value of `ratio`? Is the value of `ratio` obtained from `division.lua` and it would be `100%ph / 100%pw`?

Thank you for your valuable time.

**Comment by Omikhleia** [comment, marked as answer]
Er. LPeg is hard and I am not always sure I got it right.
This being said, here, the "+" between the expressions being a "or" alternative, I want to parse:
- `division N r` --> n = N, ratio = r (explicit values for both, i.e. the `ws * number * ws * number` case),
- or `division N` --> n = N, ratio = nil (the one-number case `ws * number`; and this, later in the logic, means we will indeed use the page ratio (`100%ph / 100%pw`) in that case),
- or `division` --> n = 9 (as a default), ratio = nil (i.e. the `lpeg.Cc(9)`  case; and same as above regarding it).

Does it make sense?

**Reply by Omikhleia** [reply]
(Slightly edited my previous answer with more details)

**Reply by no-vici** [reply]
> Er. LPeg is hard and I am not always sure I got it right. This being said, here, the "+" between the expressions being a "or" alternative ...

Yes, I'm still reading the docs and practicing. It works beautifully and obviously you've got it right. 

> * `division N r` --> n = N, ratio = r (explicit values for both, i.e. the `ws * number * ws * number` case),
> * or `division N` --> n = N, ratio = nil (the one-number case `ws * number`; and this, later in the logic, means we will indeed use the page ratio (`100%ph / 100%pw`) in that case),

Oh, now I get the logic, thank you. In the second case, if `ratio = nil`, then the default value will be used instead.

> Does it make sense?

It obviously makes sense. So, the `layoutParser` returns `division` and how the book class pick up the layout and uses it in `self:defineMaster`?

**Reply by Omikhleia** [reply]
> how the book class pick up the layout and uses it in self:defineMaster?

To be honest, I think you are too focused here on internal implementation details---that is, _how_ it works (rather than _does_ it work and solve a problem in an interesting way). It's not that I don't want to answer that question, but... had you asked a few months earlier, before #16 made it, the answer could have been quite different, regarding the "how". For end-users, it would not have been _much_ different actually. But the internal implementation changed a lot, to become more robust and extensible. And it might change again... Resilient v2.0 aims were to complete a number of tasks to lay the ground for more advanced features and make it easier for end-users, but if need be, the internal implementation _will_ change (notwithstanding the need for SILE itself to fix ancient issues and address them appropriately). My focus is on end-user experience (that is, cool tools for producing books), however implemented _deep inside_.

I will note that SILE's documentation on class & package design is being improved (probably in SILE 0.14.10, the changes to the manual just got merged into the master branch recently). At some point, I might also document here how to implement resilient-compatible modules on top of that... When it stabilizes enough. Hey, that would also mean it came to a kind of endpoint in its development, I'm not even sure this is a good perspective ;-)

**Reply by no-vici** [reply]
I appreciate "That is how it works," with the knowledge of that the underneath mechanism changes all the time. 

The reason I asked was that I was struggling to change the ratio for the division layout as the default ratio leaves a bit of space on A4, and it is still not possible to alter it with SILE's code. The last resource is too dive into Lua code. I have managed to change the value of `n` and swap the odd and even frame to leave more room for binding the book without swallowing away some text as shown in the screenshot.
![image](https://user-images.githubusercontent.com/113570767/233882347-c756f6bb-8384-423f-bde5-3885fbe4fc53.png)

I was amazed with your code and the philosophy behind it and became curious as well. Please don't blame for focusing too much on the code beneath -:)

I know that it isn't easy to explain to others how things work if they pick them up from the middle. It obviously took you a long time and a lot of hard work to be at this stage.

I have to nibble the code slowly myself for my own interest though ;D)

Highly appreciate your response as always.

**Reply by no-vici** [reply]
> > how the book class pick up the layout and uses it in self:defineMaster?

I'm slowly getting it. At first, I did not understand the logic of the `layoutParse`, so I dived into it and changed `n` in `lpeg.Cc(n)`. After seeing `P':'`, I can change the ratio in the class option without touching the `layoutParser`'s code. 

`layout=division:12` works like a charm, very clever! What is in your mind, genius?

I wish you will put some use cases in the future versions of the manual. They would save beginners like me some sweat :D).

Big thanks, again.

**Reply by Omikhleia** [reply]
> I wish you will put some use cases in the future versions of the manual. They would save beginners like me some sweat :D).

But isn't it documented already?

![image](https://user-images.githubusercontent.com/18075640/233935509-f688d385-6848-4684-88ab-f920726f2601.png)

(And the general `division N` is mentioned just after the illustration, so e.g. `division 8` would be possible although "non-usual" in historical layouts).

To make it clear, the parser allows `division 12` with spaces, _as well as_ `division:12` with a colon (as you noted about the `P':'`) but I did not document the latter---It was what the previous version of the code briefly implemented in a less generic way, so I just kept it for compatibility with my older documents but kept that silent.

So in my view, this is already documented. Whether the current documentation is clear enough, that's another question. Issues welcome for clarifying things if needed!

> (...)  the division layout as the default ratio leaves a bit of space on A4 and  it is still not possible to alter it with SILE's code. The last resource is too dive into Lua code.

Well there's nothing that prevent you from using the existing SILE ways for defining custom page masters early in a document with appropriate commands---with, for instance the **masters** package and `\define-master-template`, `\switch-master-one-page`, `\swich-master`; or with the lower-level `\pagetemplate` etc.

If one really wants "custom" page masters, that still the way to go...

The very idea of the `layout` in resilient classes is to provide a way to pick "standard" page masters (historical ones, publisher ones, etc.) rather let people use bad or poor designs. Swapping the odd and even frame and getting larger inner margin than outer margins is _usually_ quite bad, for instance (or one should consider using a gutter `offset`, also documented, to shift the layout rather than swapping it, depending on the number of pages and other considerations).

I do plan to add more predefined layouts in the future. Notably I recently had a discussion with a pocket-sized book publisher and was told what their layout was. (Interestingly, the answer was, I am simplifying, along the line of "Oh, well, we just tell the composer to use these numerical values, corresponding to a 10x17cm page. But of course the original formula can be generalized to other page sizes, I guess")---So eventually I'll add that layout.

**Reply by Omikhleia** [reply]
Re: "Swapping the odd and even frame and getting larger inner margin than outer margins is usually quite bad, for instance (or one should consider using a gutter `offset`, also documented, to shift the layout rather than swapping it, depending on the number of pages and other considerations)"

Would the added figure help here?

![image](https://user-images.githubusercontent.com/18075640/233947128-26b1ccc3-f5c4-4cc0-a5e4-8a9085938a8f.png)

Or how can I improve the documentation?

**Reply by no-vici** [reply]
> But isn't it documented already?

Thank you for pointing this out, what I meant was `layout=division 9` for the class option. I've read the manual twice but still did not figure it out. I was dump though.


> So in my view, this is already documented. Whether the current documentation is clear enough, that's another question. Issues welcome for clarifying things if needed!

MHO, an example shows how to apply a particular layout would be nice, from a beginner's point of view, Par I of the book look like an introduction theory. Don't blame me though. I will read that part again. 
> [...] Swapping the odd and even frame and getting larger inner margin than outer margins is _usually_ quite bad, for instance (or one should consider using a `gutter`, also documented, to shift the layout rather than swapping it, depending on the number of pages and other considerations).

One of my friends printed the book I'd compiled and he found it struggle to read it, because some of the text was covered by the stapled edges. He's has asked me if I could swap the mirrored pages, so I played with the layout. I will swap them back and apply the `offset` as you've suggested.

Reading along your code, I've seen the intension of standardising the layouts. I think those are for professional publications. For old men like myself, we just need something clear to read ;).

I've started with Lua for fun and to keep my brain going, now I end up learning SILE and the Resilient class. What an adventure!

**Reply by no-vici** [reply]
> Would the added figure help here?
Yep, this is very helpful! Thank you. The manual I printed a while ago did not have 1.4 section, it stops at 1.2.4 The legacy layout. So for `dimen`, could it be %pw or in pixels?

**Reply by Omikhleia** [reply]
> So for dimen, could it be %pw or in pixels?

Could be anything, but these are usually specified in inches or mm

For instance, here is a table from Lulu.com 

![image](https://user-images.githubusercontent.com/18075640/233955855-59d77cad-cd34-4e88-b5ed-44ae7518a55b.png)

I could add such a table, but the values really depend (other print-on-demand services would recommend other values, and there's also different tables depending whether the book is a hardback or a paperback, etc.) -- It's _at best_ indicative ^^ and Lulu here is very cautious (real big values IMHO). E.g. for an "ateliers demiluxe" layout, I'd likely use at most 0.25in rather than their suggested 0.5in for 300+ pages... It does seem sufficient to me, but these things are tricky.

**Reply by no-vici** [reply]
Thank you for explaining the `«dimen»` clearly. 

You are right, those recommended figures are quite tricky. I had to used smaller numbers.

**Reply by Omikhleia** [reply]
> You are right, those recommended figures are quite tricky. I had to used smaller numbers.

Yes :) The problem, I would guess, for many print-on-demand services is that they offer various paper sizes, various paper quality (thickness), and various binding methods: paperback binding, hardcover binding... and some even have coil binding---and those coils are going to eat a whole lot of the inner margins! So they won't detail all possible issues depending on all possible cases, but will just recommend pretty wide margins :rofl:

**Reply by no-vici** [reply]
>  ... and some even have coil binding---and those coils are going to eat a whole lot of the inner margins! 

Mine was ring bound, not much room left from the text body ;)

Thank you once again for your valuable time and patience.

Cheers.

**Comment by Omikhleia** [comment]
> Thank you once again for your valuable time and patience.

You are welcome! I'll probably tag a 2.0.1 release soon, so anything that can still make it in that milestone is great. I created an issue #39 to take into account the proposed clarifications on the offset.
